src/main/java/com/google/sps/servlets/AddClassTA.java

```diff
@@ -53,20 +56,48 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
           // Obtain the teaching assistant email and search for the user
           String teachingAssistantEmail = request.getParameter("taEmail").trim();
-          UserRecord userRecord = authInstance.getUserByEmail(teachingAssistantEmail);
 
-          // Find the corresponding Class Entity
+          // Find the corresponding class Key
           String classCode = request.getParameter("classCode").trim();
           Key classKey = KeyFactory.stringToKey(classCode);
-          Entity classEntity = datastore.get(txn, classKey);
 
-          // Get the list of TAs
-          ArrayList<String> listOfClassTAs = (ArrayList) classEntity.getProperty("taList");
+          // Look for the TA in the user datastore
+          PreparedQuery queryUser =
+              datastore.prepare(
+                  new Query("User")
+                      .setFilter(
+                          new FilterPredicate(
+                              "userEmail", FilterOperator.EQUAL, teachingAssistantEmail)));
 
-          // Update the class's TA list to include the new TA
-          listOfClassTAs.add(teachingAssistantEmail);
-          classEntity.setProperty("taList", listOfClassTAs);
-          datastore.put(txn, classEntity);
+          Entity user;
+
+          // If the TA user entity doesnt exist yet, create one
+          if (queryUser.countEntities() == 0) {
+
+            List<Key> taClassesList = Arrays.asList(classKey);
+
+            user = new Entity("User");
+            user.setProperty("userEmail", teachingAssistantEmail);
+            user.setProperty("registeredClasses", Collections.emptyList());
+            user.setProperty("ownedClasses", Collections.emptyList());
+            user.setProperty("taClasses", taClassesList);
+
+            datastore.put(txn, user);
+          } else {
+            // If TA user already exists, update their ta class list
+            user = queryUser.asSingleEntity();
+            List<Key> taClassesList = (List<Key>) user.getProperty("taClasses");
+
+            // Do not add a class that is already in the TA list
+            if (taClassesList.contains(classKey)) {
+              response.sendError(HttpServletResponse.SC_FORBIDDEN);
+            }
+
+            taClassesList.add(classKey);
+            user.setProperty("taClasses", taClassesList);
+
+            datastore.put(txn, user);
+          }
 
           // Redirect to the class dashboard page
           response.sendRedirect("/dashboard.html?classCode=" + classCode);
```

Comment:
Is this redirect needed? Isn't this form already on the page you're redirecting to?

Reply:
Yes, otherwise it will redirect to the servlet page